impl more traits for ptr::Alignment, add mask method

clarfonthey opened the pull request:
Changes:

* Adds `rustc_const_unstable` attributes where missing
* Makes `log2` method const
* Adds `mask` method
* Implements `Default`, which is equivalent to `Alignment::MIN`

No longer included in PR:

* Removes indirection of `AlignmentEnum` type alias (this was intentional)
* Implements `Display`, `Binary`, `Octal`, `LowerHex`, and `UpperHex` (should go through libs-api instead)
* Controversially implements `LowerExp` and `UpperExp` using `p` instead of `e` to indicate a power of 2 (also should go through libs-api)

Tracking issue for `ptr::Alignment`: #102070

clarfonthey committed:
impl more traits for ptr::Alignment, add mask method

clarfonthey commented on library/core/src/ptr/alignment.rs:
```diff
@@ -131,6 +167,91 @@ impl fmt::Debug for Alignment {
+}
+
+impl Alignment {
+    // longest string: 1p-128
```
even though no targets have 128-bit usize, I figured I might as well future-proof it

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5985813501/job/16238293251) [(plain)](https://github.com/rust-lang/rust/commit/0841b1dae3cea47add64eeb603f949edc983a733/checks/16238293251/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=clarfonthey
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_73176393-7f7e-4b6f-96f3-b2e7b50ad217
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=135a353e26278fe8d135422f9a252f66394801b6
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_73176393-7f7e-4b6f-96f3-b2e7b50ad217
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_73176393-7f7e-4b6f-96f3-b2e7b50ad217
GITHUB_TRIGGERING_ACTOR=clarfonthey
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/115249/merge
GITHUB_WORKFLOW_SHA=135a353e26278fe8d135422f9a252f66394801b6
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_18_X64=/opt/hostedtoolcache/go/1.18.10/x64
---
.........................................

failures:

---- library/core/src/ptr/alignment.rs - ptr::alignment::Alignment::mask (line 136) stdout ----
error[E0107]: associated function takes 0 generic arguments but 1 generic argument was supplied
##[error]  --> library/core/src/ptr/alignment.rs:141:20
error: doctest failed, to rerun pass `-p core --doc`
   |
8  | let one = NonNull::dangling::<u8>().as_ptr();
   |                    ^^^^^^^^------ help: remove these generics
   |                    |
---
   |
91 |     pub const fn dangling() -> Self {
   |                  ^^^^^^^^

error[E0107]: associated function takes 0 generic arguments but 1 generic argument was supplied
   |
9  | let four = NonNull::dangling::<u32>().as_ptr();
   |                     ^^^^^^^^------- help: remove these generics
   |                     |
---

error[E0699]: cannot call a method on a raw pointer with an unknown pointee type
##[error]  --> library/core/src/ptr/alignment.rs:145:17
   |
12 | assert_eq!(four.mask(Alignment::of::<u8>().mask()), four);

error[E0699]: cannot call a method on a raw pointer with an unknown pointee type
##[error]  --> library/core/src/ptr/alignment.rs:146:17
   |
   |
13 | assert_eq!(four.mask(Alignment::of::<u16>().mask()), four);

error[E0699]: cannot call a method on a raw pointer with an unknown pointee type
##[error]  --> library/core/src/ptr/alignment.rs:147:17
   |
   |
14 | assert_eq!(four.mask(Alignment::of::<u32>().mask()), four);

error[E0699]: cannot call a method on a raw pointer with an unknown pointee type
##[error]  --> library/core/src/ptr/alignment.rs:148:17
   |
   |
15 | assert_ne!(four.mask(Alignment::of::<u64>().mask()), four);

error[E0699]: cannot call a method on a raw pointer with an unknown pointee type
##[error]  --> library/core/src/ptr/alignment.rs:149:16
   |
   |
16 | assert_eq!(one.mask(Alignment::of::<u64>().mask()), one);

error[E0308]: mismatched types
##[error]   --> library/core/src/ptr/alignment.rs:152:21
    |
    |
19  | assert_eq!(any.mask(Alignment::MIN), any);
    |                ---- ^^^^^^^^^^^^^^ expected `usize`, found `Alignment`
    |                arguments to this method are incorrect
    |
note: method defined here
   --> /checkout/library/core/src/ptr/const_ptr.rs:603:12
   --> /checkout/library/core/src/ptr/const_ptr.rs:603:12
    |
603 |     pub fn mask(self, mask: usize) -> *const T {
    |            ^^^^
help: call `Into::into` on this expression to convert `std::ptr::Alignment` into `usize`
    |
19  | assert_eq!(any.mask(Alignment::MIN.into()), any);

error: aborting due to 8 previous errors

Some errors have detailed explanations: E0107, E0308, E0699.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5986153194/job/16238970101) [(plain)](https://github.com/rust-lang/rust/commit/a6546fefd7156d94bfc09b75051b6f4df741085b/checks/16238970101/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=clarfonthey
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_56528bb1-f284-410b-8fa0-b97cb33bc07b
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=e0b2dbabddd0a8c7129c8e69528ee8dcb424a581
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_56528bb1-f284-410b-8fa0-b97cb33bc07b
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_56528bb1-f284-410b-8fa0-b97cb33bc07b
GITHUB_TRIGGERING_ACTOR=clarfonthey
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/115249/merge
GITHUB_WORKFLOW_SHA=e0b2dbabddd0a8c7129c8e69528ee8dcb424a581
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_18_X64=/opt/hostedtoolcache/go/1.18.10/x64
---
.........................................

failures:

---- library/core/src/ptr/alignment.rs - ptr::alignment::Alignment::mask (line 136) stdout ----
error[E0606]: casting `&i32` as `*const u32` is invalid
   |
   |
10 | let any = &1 as *const u32;

error: aborting due to previous error
error: aborting due to previous error
error: doctest failed, to rerun pass `-p core --doc`

For more information about this error, try `rustc --explain E0606`.
Couldn't compile the test.

```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5986597491/job/16239917300) [(plain)](https://github.com/rust-lang/rust/commit/e86b2ed786632bdd9cbf70ed93eb71f2e7e40eed/checks/16239917300/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=clarfonthey
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_81a822ff-5894-4e26-9205-21ef2308f734
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=1c152812e61154ee76c11d3d60921922a75accc1
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_81a822ff-5894-4e26-9205-21ef2308f734
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_81a822ff-5894-4e26-9205-21ef2308f734
GITHUB_TRIGGERING_ACTOR=clarfonthey
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/115249/merge
GITHUB_WORKFLOW_SHA=1c152812e61154ee76c11d3d60921922a75accc1
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_18_X64=/opt/hostedtoolcache/go/1.18.10/x64
---
   0:     0x556dd2132d6c - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h9ee22e58c91ca038
   1:     0x556dd2153b77 - core::fmt::write::hb78ce2d8451e80c4
   2:     0x556dd2130aa5 - std::io::Write::write_fmt::h9498440495c6d3f8
   3:     0x556dd2132b54 - std::sys_common::backtrace::print::h1e931e4b71f866dd
   4:     0x556dd213453f - std::panicking::panic_hook_with_disk_dump::{{closure}}::he6bf5f49228e21f2
   5:     0x556dd2134266 - std::panicking::panic_hook_with_disk_dump::hf45032a36c518821
error: doctest failed, to rerun pass `-p core --doc`
   7:     0x556dd2134977 - std::panicking::begin_panic_handler::{{closure}}::h5a03ba7f37b7faa1
   8:     0x556dd2133336 - std::sys_common::backtrace::__rust_end_short_backtrace::hec0de78d8288640b
   9:     0x556dd2134682 - rust_begin_unwind
  10:     0x556dd210e7d3 - core::panicking::panic_fmt::h3370a34ad867dd0c
  10:     0x556dd210e7d3 - core::panicking::panic_fmt::h3370a34ad867dd0c
  11:     0x556dd210eb2b - core::panicking::assert_failed_inner::h7b1e34a776646f49
  12:     0x556dd210f36a - core::panicking::assert_failed::h75934836cebf0aab
  13:     0x556dd210f9ce - rust_out::main::_doctest_main_library_core_src_ptr_alignment_rs_136_0::hf47e5a4a1901265f
  14:     0x556dd210f426 - rust_out::main::hd406d7fa8f60c984
  16:     0x556dd210f036 - std::sys_common::backtrace::__rust_begin_short_backtrace::hc8e3e74b793a632e
  17:     0x556dd210f089 - std::rt::lang_start::{{closure}}::h37744c6fcb148f93
  18:     0x556dd212e53c - std::rt::lang_start_internal::h77f2b788a85d1cb4
  19:     0x556dd210f067 - std::rt::lang_start::h1eb6fa72167ade40
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5988449834/job/16243645899) [(plain)](https://github.com/rust-lang/rust/commit/dedf2f2bf6c07a9bcbf5eadff6eacab489a37953/checks/16243645899/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=clarfonthey
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_4157dadb-c4f0-4f55-8fd6-fbedc941280f
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=e5b2056fb604e559ec15c56f0fd8e7e1de748606
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_4157dadb-c4f0-4f55-8fd6-fbedc941280f
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_4157dadb-c4f0-4f55-8fd6-fbedc941280f
GITHUB_TRIGGERING_ACTOR=clarfonthey
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/115249/merge
GITHUB_WORKFLOW_SHA=e5b2056fb604e559ec15c56f0fd8e7e1de748606
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_18_X64=/opt/hostedtoolcache/go/1.18.10/x64
---

stderr:
thread 'main' panicked at library/core/src/ptr/alignment.rs:15:1:
assertion `left == right` failed
  left: 0x0
error: doctest failed, to rerun pass `-p core --doc`
 right: 0x1
   0:     0x5612f24fabcc - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h9ee22e58c91ca038
   1:     0x5612f251b9d7 - core::fmt::write::hb78ce2d8451e80c4
   2:     0x5612f24f8905 - std::io::Write::write_fmt::h9498440495c6d3f8
   3:     0x5612f24fa9b4 - std::sys_common::backtrace::print::h1e931e4b71f866dd
   3:     0x5612f24fa9b4 - std::sys_common::backtrace::print::h1e931e4b71f866dd
   4:     0x5612f24fc39f - std::panicking::panic_hook_with_disk_dump::{{closure}}::he6bf5f49228e21f2
   5:     0x5612f24fc0c6 - std::panicking::panic_hook_with_disk_dump::hf45032a36c518821
   7:     0x5612f24fc7d7 - std::panicking::begin_panic_handler::{{closure}}::h5a03ba7f37b7faa1
   8:     0x5612f24fb196 - std::sys_common::backtrace::__rust_end_short_backtrace::hec0de78d8288640b
   9:     0x5612f24fc4e2 - rust_begin_unwind
  10:     0x5612f24d67d3 - core::panicking::panic_fmt::h3370a34ad867dd0c
  10:     0x5612f24d67d3 - core::panicking::panic_fmt::h3370a34ad867dd0c
  11:     0x5612f24d6b2b - core::panicking::assert_failed_inner::h7b1e34a776646f49
  12:     0x5612f24d730a - core::panicking::assert_failed::h75934836cebf0aab
  13:     0x5612f24d787e - rust_out::main::_doctest_main_library_core_src_ptr_alignment_rs_136_0::hf47e5a4a1901265f
  14:     0x5612f24d73a6 - rust_out::main::hd406d7fa8f60c984
  16:     0x5612f24d7036 - std::sys_common::backtrace::__rust_begin_short_backtrace::hc8e3e74b793a632e
  17:     0x5612f24d7089 - std::rt::lang_start::{{closure}}::h37744c6fcb148f93
  18:     0x5612f24f639c - std::rt::lang_start_internal::h77f2b788a85d1cb4
  19:     0x5612f24d7067 - std::rt::lang_start::h1eb6fa72167ade40
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5988579872/job/16243903904) [(plain)](https://github.com/rust-lang/rust/commit/a3c430d836d26fc273d07fc5e3b81d030f6a6f30/checks/16243903904/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=clarfonthey
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_00edd3e9-7703-4d74-ac28-084c3083a954
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=592baf2951d1f3120c78ab9218f60813f61fd373
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_00edd3e9-7703-4d74-ac28-084c3083a954
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_00edd3e9-7703-4d74-ac28-084c3083a954
GITHUB_TRIGGERING_ACTOR=clarfonthey
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/115249/merge
GITHUB_WORKFLOW_SHA=592baf2951d1f3120c78ab9218f60813f61fd373
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_18_X64=/opt/hostedtoolcache/go/1.18.10/x64
---
16              usize
17              u8
-            and 20 others
+            and 21 others
19    = note: required for `&str` to implement `UpperHex`
20 note: required by a bound in `core::fmt::rt::Argument::<'a>::new_upper_hex`
21   --> $SRC_DIR/core/src/fmt/rt.rs:LL:COL

The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/fmt/ifmt-unimpl/ifmt-unimpl.stderr
To update references, rerun the tests and pass the `--bless` flag
---
--- stderr -------------------------------
error[E0277]: the trait bound `str: UpperHex` is not satisfied
##[error]  --> /checkout/tests/ui/fmt/ifmt-unimpl.rs:2:21
   |
LL |     format!("{:X}", "3");
   |              ----   ^^^ the trait `UpperHex` is not implemented for `str`
   |              required by a bound introduced by this call
   |
   |
   = help: the following other types implement trait `UpperHex`:
             i8
             i16
             i32
             i64
             i64
             i128
             usize
             u8
           and 21 others
   = note: required for `&str` to implement `UpperHex`
note: required by a bound in `core::fmt::rt::Argument::<'a>::new_upper_hex`
  --> /rustc/FAKE_PREFIX/library/core/src/fmt/rt.rs:112:5

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
```

</details>

scottmcm commented on library/core/src/ptr/alignment.rs:
```diff
@@ -118,10 +121,39 @@ impl Alignment {
+    #[inline]
+    pub const fn mask(self) -> usize {
+        // SAFETY: The alignment is always nonzero, and therefore decrementing won't overflow.
+        !(unsafe { self.as_usize().unchecked_sub(1) })
```
Sadly, `unchecked_sub(1)` is useless on unsigned types <https://github.com/llvm/llvm-project/issues/53377>

I guess it's fine to leave it, though, since it's clearly correct.  And trying to do it in `isize` would be UB.

clarfonthey replied:
Oh, that's unfortunate. I even was considering at some point adding a `decrement()` method to `NonZeroU*` that could use `unchecked_sub` since it would never overflow, but I guess that that also wouldn't work.

scottmcm commented on library/core/src/ptr/alignment.rs:
```diff
@@ -131,6 +163,92 @@ impl fmt::Debug for Alignment {
+#[unstable(feature = "ptr_alignment_type", issue = "102070")]
+impl fmt::Display for Alignment {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(&self.as_nonzero(), f)
```
Pondering: I have no idea when we should be including `Display` on types.  I guess it's not a problem, but it feels like alignment isn't something I'd usually be showing to people.  But showing it in binary or hex seems logical, so maybe display is ok too.

clarfonthey replied:
Yeah, that was my thought too. It's less that it's necessarily a common operation, but more that it's well-defined and not a whole lot of extra code to do. I'm fine removing it if it seems too weird.

scottmcm commented on library/core/src/ptr/alignment.rs:
```diff
@@ -193,16 +311,17 @@ impl hash::Hash for Alignment {
     }
 }
 
-#[cfg(target_pointer_width = "16")]
```
I was specifically asked to do this this way -- https://github.com/rust-lang/rust/pull/95361#discussion_r846530028 -- so unless you feel particularly strongly, I'd say to just leave it like this.

(By having `AlignmentEnum16` not cfg'd out it means you know it actually compiles without checking on a 16-bit target.)

clarfonthey replied:
I'm fine reverting this change tbqh

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5992963838/job/16252930098) [(plain)](https://github.com/rust-lang/rust/commit/f8cd3163d61f39a3dbf70e95e206f4a68b683865/checks/16252930098/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=clarfonthey
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_8cf8f71a-0996-45b9-a4d4-68ce87cb61d8
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=19a8453dc0c9e7487d6cd052ae11a920488ad3ed
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_8cf8f71a-0996-45b9-a4d4-68ce87cb61d8
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_8cf8f71a-0996-45b9-a4d4-68ce87cb61d8
GITHUB_TRIGGERING_ACTOR=clarfonthey
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/115249/merge
GITHUB_WORKFLOW_SHA=19a8453dc0c9e7487d6cd052ae11a920488ad3ed
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_18_X64=/opt/hostedtoolcache/go/1.18.10/x64
---
...................i........................i........................................... 4136/4177
.........................................

failures:
error: doctest failed, to rerun pass `-p core --doc`
---- library/core/src/ptr/alignment.rs - ptr::alignment::Alignment::mask (line 136) stdout ----
Test executable failed (exit status: 101).

stderr:
---
   0:     0x55938931bbcc - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h9ee22e58c91ca038
   1:     0x55938933c9d7 - core::fmt::write::hb78ce2d8451e80c4
   2:     0x559389319905 - std::io::Write::write_fmt::h9498440495c6d3f8
   3:     0x55938931b9b4 - std::sys_common::backtrace::print::h1e931e4b71f866dd
   4:     0x55938931d39f - std::panicking::panic_hook_with_disk_dump::{{closure}}::he6bf5f49228e21f2
   5:     0x55938931d0c6 - std::panicking::panic_hook_with_disk_dump::hf45032a36c518821
   7:     0x55938931d7d7 - std::panicking::begin_panic_handler::{{closure}}::h5a03ba7f37b7faa1
   8:     0x55938931c196 - std::sys_common::backtrace::__rust_end_short_backtrace::hec0de78d8288640b
   9:     0x55938931d4e2 - rust_begin_unwind
  10:     0x5593892f77d3 - core::panicking::panic_fmt::h3370a34ad867dd0c
  10:     0x5593892f77d3 - core::panicking::panic_fmt::h3370a34ad867dd0c
  11:     0x5593892f7b2b - core::panicking::assert_failed_inner::h7b1e34a776646f49
  12:     0x5593892f830a - core::panicking::assert_failed::h75934836cebf0aab
  13:     0x5593892f887e - rust_out::main::_doctest_main_library_core_src_ptr_alignment_rs_136_0::hf47e5a4a1901265f
  14:     0x5593892f83a6 - rust_out::main::hd406d7fa8f60c984
  16:     0x5593892f8036 - std::sys_common::backtrace::__rust_begin_short_backtrace::hc8e3e74b793a632e
  17:     0x5593892f8089 - std::rt::lang_start::{{closure}}::h37744c6fcb148f93
  18:     0x55938931739c - std::rt::lang_start_internal::h77f2b788a85d1cb4
  19:     0x5593892f8067 - std::rt::lang_start::h1eb6fa72167ade40
```

</details>

clarfonthey commented:
Removed the formatting impls and added a comment to the `Default` impl. If there's anything else you'd like me to do, @scottmcm, please let me know.

@rustbot ready

bors commented:
:pushpin: Commit 5dda0c99b4c70364af55744f3cda1b139c78ef7f has been approved by `scottmcm`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=scottmcm 5dda0c99b4c70364af55744f3cda1b139c78ef7f -->
<!-- homu: {"type":"Approved","sha":"5dda0c99b4c70364af55744f3cda1b139c78ef7f","approver":"scottmcm","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 5dda0c99b4c70364af55744f3cda1b139c78ef7f with merge 2c4b73c7b57e8d782a10578a5f1256510baab917...
<!-- homu: {"type":"BuildStarted","head_sha":"5dda0c99b4c70364af55744f3cda1b139c78ef7f","merge_sha":"2c4b73c7b57e8d782a10578a5f1256510baab917"} -->

rust-log-analyzer commented:
The job **`i686-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/6514190367/job/17695038750) [(plain)](https://github.com/rust-lang-ci/rust/commit/2c4b73c7b57e8d782a10578a5f1256510baab917/checks/17695038750/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6514190367/job/17695869234)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/6514190367/job/17695869234","builder_name":"checks-actions"} -->

clarfonthey commented:
I don't like it, but I updated the example to have explicitly-aligned structs.

Right now, it's operating way more as a test than an example of behaviour, but that can be resolved during stabilisation.

clarfonthey commented:
Ping @scottmcm when you have time to take a look at this

clarfonthey commented:
Ping @scottmcm again when you're free.

bors commented:
:pushpin: Commit 114873dc19d2c46f0090a7485218b1fb202648fc has been approved by `scottmcm`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=scottmcm 114873dc19d2c46f0090a7485218b1fb202648fc -->
<!-- homu: {"type":"Approved","sha":"114873dc19d2c46f0090a7485218b1fb202648fc","approver":"scottmcm","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 114873dc19d2c46f0090a7485218b1fb202648fc with merge 61d3b263a793b390f6231f08d862e8c71d04e3ef...
<!-- homu: {"type":"BuildStarted","head_sha":"114873dc19d2c46f0090a7485218b1fb202648fc","merge_sha":"61d3b263a793b390f6231f08d862e8c71d04e3ef"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6912440337/job/18809246316)
Approved by: scottmcm
Pushing 61d3b263a793b390f6231f08d862e8c71d04e3ef to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"scottmcm","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6912440337/job/18809246316"},"merge_sha":"61d3b263a793b390f6231f08d862e8c71d04e3ef"} -->

rust-timer commented:
Finished benchmarking commit (61d3b263a793b390f6231f08d862e8c71d04e3ef): [comparison URL](https://perf.rust-lang.org/compare.html?start=e6dade96f4f16b2c0f621d657130b90b4bb519a2&end=61d3b263a793b390f6231f08d862e8c71d04e3ef&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=e6dade96f4f16b2c0f621d657130b90b4bb519a2&end=61d3b263a793b390f6231f08d862e8c71d04e3ef&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=e6dade96f4f16b2c0f621d657130b90b4bb519a2&end=61d3b263a793b390f6231f08d862e8c71d04e3ef&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.9% | [2.4%, 4.9%] | 6     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=e6dade96f4f16b2c0f621d657130b90b4bb519a2&end=61d3b263a793b390f6231f08d862e8c71d04e3ef&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.4% | [1.8%, 2.9%] | 10    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=e6dade96f4f16b2c0f621d657130b90b4bb519a2&end=61d3b263a793b390f6231f08d862e8c71d04e3ef&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 675.975s -> 678.192s (0.33%)
**Artifact size**: 313.82 MiB -> 313.84 MiB (0.01%)
